sound/soc/sof/core.c

```diff
@@ -353,24 +339,28 @@ int snd_sof_device_remove(struct device *dev)
 	if (IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE))
 		cancel_work_sync(&sdev->probe_work);
 
-	snd_sof_fw_unload(sdev);
-	snd_sof_ipc_free(sdev);
-	snd_sof_free_debug(sdev);
-	snd_sof_free_trace(sdev);
+	if (sdev->fw_state > SOF_FW_BOOT_NOT_STARTED) {
```

Comment:
And this check should be for > PREPARE

Reply:
@ranj063 Hmm. But if e.g. PM resume fails to load FW, state will be BOOT_PREPARE. If the check would be >PREPARE here, that would mean we would not free ipc, debug and trace, which is wrong. We should free the resources also in this case. Only case where we should not free the resources if the initial sof_probe_continue failed and we cleaned up everything. In all other substates, the resources are still allocated so we should do normal remove for the PCI device.